File: app/src/main/java/de/bikebean/app/db/state/StateFactory.java
package de.bikebean.app.db.state;

import androidx.annotation.NonNull;

import de.bikebean.app.db.sms.Sms;

public abstract class StateFactory {

    public static @NonNull State createStateWithEnum(
            long timestamp,
            final @NonNull State.KEY key,
            final @NonNull Double value,
            final @NonNull String longValue,
            final @NonNull State.STATUS state,
            int smsId
    ) {
        return new State(
                timestamp,
                key.get(),
                value,
                longValue,
                state.ordinal(),
                smsId
        );
    }

    public static @NonNull State createSimplePendingState(
            final @NonNull Sms sms,
            final @NonNull State.KEY key,
            final double value
    ) {
        return new State(
                sms.getTimestamp(),
                key.get(),
                value,
                "",
                State.STATUS.PENDING.ordinal(),
                sms.getId()
        );
    }

    public static @NonNull State createPendingState(
            final @NonNull State.KEY key,
            double value
    ) {
        return new State(
                System.currentTimeMillis(),
                key.get(),
                value,
                "",
                State.STATUS.PENDING.ordinal(),
                0
        );
    }

    public static @NonNull State createNumberState(
            final @NonNull Sms sms,
            final @NonNull State.KEY key,
            double value,
            final @NonNull State.STATUS status
    ) {
        return new State(
                sms.getTimestamp(),
                key.get(),
                value,
                "",
                status.ordinal(),
                sms.getId()
        );
    }

    public static @NonNull State createStringState(
            final @NonNull Sms sms,
            final @NonNull State.KEY key,
            final @NonNull String string,
            final @NonNull State.STATUS status
    ) {
        return new State(
                sms.getTimestamp(),
                key.get(),
                0.0,
                string,
                status.ordinal(),
                sms.getId()
        );
    }

    public static @NonNull State createNullState() {
        return new State(
            0,
            "",
            0.0,
            "",
            0,
            0
        );
    }
}

File: app/src/main/java/de/bikebean/app/db/state/StateFactory.kt
package de.bikebean.app.db.state

import de.bikebean.app.db.settings.settings.add_to_list_settings.WappState
import de.bikebean.app.db.sms.Sms
import de.bikebean.app.db.sms.SmsFactory
import de.bikebean.app.db.state.State.KEY

object StateFactory {

    fun createStateFromWappState(
            wappState: WappState,
            key: KEY,
            value: Double
    ): State {
        return State(
                wappState.sms.timestamp + 1,
                key.get(),
                value,
                "",
                State.STATUS.CONFIRMED.ordinal,
                wappState.sms.id
        )
    }

    fun createSimplePendingState(
            sms: Sms,
            key: KEY,
            value: Double
    ): State = State(
            sms.timestamp,
            key.get(),
            value,
            "",
            State.STATUS.PENDING.ordinal,
            sms.id
    )

    @JvmStatic
    fun createPendingState(
            key: KEY,
            value: Double
    ): State = State(
            System.currentTimeMillis(),
            key.get(),
            value,
            "",
            State.STATUS.PENDING.ordinal,
            0
    )

    fun createNumberState(
            sms: Sms,
            key: KEY,
            value: Double,
            status: State.STATUS
    ): State = State(
            sms.timestamp,
            key.get(),
            value,
            "",
            status.ordinal,
            sms.id
    )

    fun createUnsetState(
            key: KEY,
            value: Double
    ): State = SmsFactory.createNullSms().let {
        State(
                it.timestamp,
                key.get(),
                value,
                "",
                State.STATUS.UNSET.ordinal,
                it.id
        )
    }

    fun createStringState(
            sms: Sms,
            key: KEY,
            string: String,
            status: State.STATUS
    ): State = State(
            sms.timestamp,
            key.get(),
            0.0,
            string,
            status.ordinal,
            sms.id
    )

    @JvmStatic
    fun createNullState(): State = State(
            0,
            "",
            0.0,
            "",
            0,
            0
    )
}